spellcheck/platform/linux/spellcheck_linux.cpp

```diff
@@ -1,38 +1,117 @@
 // This file is part of Desktop App Toolkit,
 // a set of libraries for developing nice desktop applications.
 //
-// For license and copyright information please follow this link:
-// https://github.com/desktop-app/legal/blob/master/LEGAL
-//
+// Author: Nicholas Guriev <guriev-ns@ya.ru>, public domain, 2019
+// License: CC0, https://creativecommons.org/publicdomain/zero/1.0/legalcode
+
+#include <set>
+#include "spellcheck/spellcheck_utils.h"
+#include "spellcheck/platform/linux/enchant.h"
+
 #include "spellcheck/platform/linux/spellcheck_linux.h"
 
-namespace Platform::Spellchecker {
+namespace {
+
+std::unique_ptr<enchant::Broker> brokerHandle;
+std::vector<std::unique_ptr<enchant::Dict>> validators;
+enchant::Dict *personal;
 
-bool CheckSpelling(const QString &wordToCheck) {
-	return true;
+void GetValidators() {
+	static bool tried = false;
+	if (tried) return;
+	tried = true;
+	if (!enchant::loader::do_explicit_linking()) return;
+
+	typedef std::unique_ptr<enchant::Dict> DictPtr;
+	brokerHandle = std::make_unique<enchant::Broker>();
+	personal = brokerHandle->request_pwl_dict("personal.dic");
+	validators.push_back(DictPtr(personal));
+
+	std::set<std::string> langs;
+	brokerHandle->list_dicts([](
+			const char *language,
+			const char *provider,
+			const char *description,
+			const char *filename,
+			void *our_payload) {
+		static_cast<decltype(langs)*>(our_payload)->insert(language);
+	}, &langs);
+	for (const auto &language : langs) {
+		validators.push_back(DictPtr(brokerHandle->request_dict(language)));
+	}
 }
 
-void FillSuggestionList(
-	const QString &wrongWord,
-	std::vector<QString> *optionalSuggestions) {
+}  // anonymous namespace
+
+bool Platform::Spellchecker::IsAvailable() {
+	GetValidators();
+	return validators.size() > 1;
 }
 
-void AddWord(const QString &word) {
+void Platform::Spellchecker::KnownLanguages(std::vector<QString> *langCodes) {
+	GetValidators();
+	langCodes->clear();
+	for (const auto &validator : validators) {
+		if (validator.get() == personal) continue;
+		langCodes->push_back(validator->get_lang().c_str());
+	}
 }
 
-void RemoveWord(const QString &word) {
+bool Platform::Spellchecker::CheckSpelling(const QString &wordToCheck) {
+	GetValidators();
+
+	return ranges::any_of(validators, [&](const auto &validator) {
+		return validator->check(wordToCheck.toStdString());
+	}) || validators.empty();
 }
 
-void IgnoreWord(const QString &word) {
+void Platform::Spellchecker::FillSuggestionList(
+		const QString &wrongWord,
+		std::vector<QString> *optionalSuggestions) {
+	GetValidators();
+
+	optionalSuggestions->clear();
+	for (const auto &validator : validators) {
+		auto suggestions = validator->suggest(wrongWord.toStdString());
```

Comment:
Nitpick: please move `wrongWord.toStdString()` out of scope and place it to the `const auto` variable.